src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyGraph.cs

```diff
@@ -143,14 +190,181 @@ public class DependencyGraph
+                    // a build, then no diff from latest.
+                    if (newestBuildWithChannel != null)
+                    {
+                        int channelId = newestBuildWithChannel.Channels.First().Id.Value;
+                        // Just choose the first channel. This algorithm is mostly just heuristic.
+                        string latestCommitKey = $"{node.RepoUri}@{channelId}";
+                        string latestCommit = null;
+                        if (!latestCommitCache.TryGetValue(latestCommitKey, out latestCommit))
+                        {
+                            // Look up latest build in the channel
+                            var latestBuild = await barOnlyRemote.GetLatestBuildAsync(node.RepoUri, channelId);
+                            // Could be null, if the only build was removed from the channel
+                            if (latestBuild != null)
+                            {
+                                latestCommit = latestBuild.Commit;
+                            }
+                            // Add to cache
+                            latestCommitCache.Add(latestCommitKey, latestCommit);
+                        }
+
+                        // Perform diff if there is a latest commit.
+                        if (!string.IsNullOrEmpty(latestCommit))
+                        {
+                            IRemote repoRemote = remoteFactory.GetRemote(node.RepoUri, logger);
+                            // This will return a no-diff if latestCommit == node.Commit
+                            node.DiffFrom = await repoRemote.GitDiffAsync(node.RepoUri, latestCommit, node.Commit);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Diff each node in the graph against the latest build in
+        ///     the graph.
+        /// </summary>
+        /// <param name="remoteFactory"></param>
+        /// <param name="logger"></param>
+        /// <param name="nodeCache"></param>
+        /// <param name="visitedRepoUriNodes"></param>
+        /// <returns></returns>
+        private static async Task DoLatestInGraphNodeDiffAsync(
+            IRemoteFactory remoteFactory,
+            ILogger logger,
+            Dictionary<string, DependencyGraphNode> nodeCache,
+            Dictionary<string, DependencyGraphNode> visitedRepoUriNodes)
+        {
+            logger.LogInformation("Running latest in graph node diff.");
+
+            // Find the build of each repo in the graph, then
+            // get the diff info from the latest
+            Dictionary<string, DependencyGraphNode>.KeyCollection uniqueRepos = visitedRepoUriNodes.Keys;
+            foreach (string repo in uniqueRepos)
+            {
+                // Get all nodes with this value
+                IEnumerable<DependencyGraphNode> nodes = nodeCache.Values.Where(n => n.RepoUri == repo);
```

Comment:
>nodeCache [](start = 57, length = 9)

Why does this take a dictionary... but then never use that dictionary? it should just take an IEnumerable<DependencyGraphNode> and use that.

Reply:
The other parameter we only use the keys of. It's sort of a weird function.  We are using two halves of two dictionaries...

---
In reply to: [261782723](https://github.com/dotnet/arcade-services/pull/170#discussion_r261782723) [](ancestors = 261782723)

Reply:
In this particular piece of code, it's probably not a problem as it stands today, but each call to these methods needs to fully enumerate the IEnumerable.  That can have a lot of strange consequences depending on what that enumerable is.  It will rerun the Where clauses, if it was generated by an method with "yield return" statements, it will restart the method execution from it's initial state, which can often have very unexpected consequences.  It might return a totally different set of data if the generation mechanism isn't stable.

Rarely does any implementation of IEnumerable really guard against multiple enumeration in this manner.  So to be safe, and make sure none of that is happening, if you need to muck about with the same enumeration more than once, call the ToList() method to make it a nice, unchanging, not-reentrant List<T>.

Like I said, I was able to trace the source of this enumeration so it's not wrong _right now_... but someone could easily refactor the callers without realizing this method is doing this, and it might be wrong in the future, and it's easy to guard against it now.

Reply:
I gotcha.  Sounds fine to me.  The primary issue i'm worried about here is avoiding over-copying collections. Do you happen to know what the performance characteristcs of ToList() are?

Reply:
In my reading of the code, these collections seemed small to me... A few hundred items or less. So the characteristics of making a tiny list of pointers should be negligible.  I feel like it was based on some "Where" clause from some other list, so the cost of going to a list once is probably much better than enumerating the list 3 times and reevaluating all the where clauses.  Especially since the first call is "Count()"... which does a full enumeration, which the list would remove entirely (since it has a built in count).